pkg/tracing/client/observability_service.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+
+	obsclient "github.com/cloudevents/sdk-go/observability/opencensus/v2/client"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/client"
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/cloudevents/sdk-go/v2/observability"
+	"go.opencensus.io/trace"
+
+	"knative.dev/eventing/pkg/tracing"
+)
+
+func New() knativeObservabilityService {
+	return knativeObservabilityService{
+		obsclient.New(),
+	}
+}
+
+type knativeObservabilityService struct {
+	base client.ObservabilityService
+}
+
+func (k knativeObservabilityService) InboundContextDecorators() []func(context.Context, binding.Message) context.Context {
+	return k.base.InboundContextDecorators()
+}
+
+func (k knativeObservabilityService) RecordReceivedMalformedEvent(ctx context.Context, err error) {
+	k.base.RecordReceivedMalformedEvent(ctx, err)
+}
+
+func (k knativeObservabilityService) RecordCallingInvoker(ctx context.Context, event *event.Event) (context.Context, func(errOrResult error)) {
+	return k.base.RecordCallingInvoker(ctx, event)
+}
+
+func (k knativeObservabilityService) RecordSendingEvent(ctx context.Context, event event.Event) (context.Context, func(errOrResult error)) {
+	// From https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#span-name
+	// The span name SHOULD be set to the message destination name and the operation being performed
+	spanName := observability.ClientSpanName
+	spanKind := trace.SpanKindClient
+
+	spanData := tracing.SpanDataFromContext(ctx)
+	if spanData != nil {
+		spanName = spanData.Name
+		spanKind = spanData.Kind
+	}
+
+	ctx, span := trace.StartSpan(ctx, spanName, trace.WithSpanKind(spanKind))
+	span.AddAttributes(obsclient.EventTraceAttributes(&event)...)
+	if spanData != nil && len(spanData.Attributes) > 0 {
+		span.AddAttributes(spanData.Attributes...)
+	}
+
+	return ctx, func(errOrResult error) {
+		span.End()
+	}
```

Comment:
The CE SDK implementation seems to bring a valuable tag to the span:

A result reporter is created per span:

https://github.com/cloudevents/sdk-go/blob/1170e89edb9b504a806f2c6a26563c3c26b68276/observability/opencensus/v2/client/observability_service.go#L44

And deferred to check the outcome, depending on it a tag is added:

https://github.com/cloudevents/sdk-go/blob/1170e89edb9b504a806f2c6a26563c3c26b68276/observability/opencensus/v2/client/observability_service.go#L50-L57

https://github.com/cloudevents/sdk-go/blob/1170e89edb9b504a806f2c6a26563c3c26b68276/observability/opencensus/v2/client/reporter.go#L77-L84

Do you think something like that could be added to this PR's implementation?

Reply:
Thanks for bringing this up. The reason I removed the reporter is because it's about metrics, not traces, and I have on my TODO list to check if this is not redundant with [stats reporter](https://github.com/knative/eventing/blob/main/pkg/metrics/source/stats_reporter.go#L30)... and it's not. We need to have it for not breaking the existing behavior (though we don't document it)